### PR TITLE
service/s3: Add documentation for using unseekable body PutObject and UploadPart 

### PR DESCRIPTION
**Pull request by jasdel**
Adds documentation to the PutObject and UploadPart operations Body member how to upload unseekable objects to an Amazon S3 Bucket.

Fixes: #1108
Depends on: #1175

### CODE REVIEW

File: .changes/next-release/service.s3-bugfix-1615945295096775000.json [hunk 1]
```diff
@@ -0,0 +1,9 @@
+{
+ "ID": "service.s3-bugfix-1615945295096775000",
+ "SchemaVersion": 1,
+ "Module": "service/s3",
+ "Type": "bugfix",
```
Comment (skotambkar) [inline .changes/next-release/service.s3-bugfix-1615945295096775000.json:5]: Is this bugFix type intentional?

File: aws/signer/v4/middleware.go [hunk 2]
```diff
@@ -156,6 +156,16 @@ func (m *computePayloadSHA256) HandleBuild(
...
 
+// SwapComputePayloadSHA256ForUnsignedPayloadMiddleware replaces the
+// ComputePayloadSHA256 middleware with the UnsignedPayload middleware.
+//
+// Use this to disable computing the Payload SHA256 checksum and instead use
+// UNSIGNED-PAYLOAD for the SHA256 value.
+func SwapComputePayloadSHA256ForUnsignedPayloadMiddleware(stack *middleware.Stack) error {
+	_, err := stack.Build.Swap(computePayloadHashMiddlewareID, &unsignedPayload{})
+	return err
+}
```
Comment (skotambkar) [inline aws/signer/v4/middleware.go:167]: Don't think we should export this as standard util in the SDK. Unsigned payload doesn't work with all services, and is not a security best practice. Why is this required?

File: codegen/smithy-aws-go-codegen/src/main/java/software/amazon/smithy/aws/go/codegen/customization/S3AddPutObjectUnseekableBodyDoc.java [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+            ShapeId.from("com.amazonaws.s3#AmazonS3"), SetUtils.of(
+                    new Pair(ShapeId.from("com.amazonaws.s3#PutObjectRequest"), "Body"),
+                    new Pair(ShapeId.from("com.amazonaws.s3#UploadPartRequest"), "Body")
+            )
+    );
+
+    @Override
+    public byte getOrder() {
+        // This integration should happen before other integrations that rely on the presence of this trait
+        return -60;
```
Comment (skotambkar) [inline codegen/smithy-aws-go-codegen/src/main/java/software/amazon/smithy/aws/go/codegen/customization/S3AddPutObjectUnseekableBodyDoc.java:33]: Wouldn't pre-process model always run prior to usage of model? Don't think this ordering is required.

Comment (jasdel) [reply codegen/smithy-aws-go-codegen/src/main/java/software/amazon/smithy/aws/go/codegen/customization/S3AddPutObjectUnseekableBodyDoc.java:33]: Can remove. The ordering is always present, will inherit from parent if not defined here. But yeah I don't think order maters in this case. Other customizations where order model is pre-processed is where this will be relevant.

File: .changes/next-release/service.s3-bugfix-1615945295096775000.json (hunk 1 above)
Comment (jasdel) [reply .changes/next-release/service.s3-bugfix-1615945295096775000.json:5]: Yeah, the changetool doesn't have `documentation` type, `announcement` doesn't really fit well, but could use that instead of `bugfix`.

File: aws/signer/v4/middleware.go (hunk 2 above)
Comment (jasdel) [reply aws/signer/v4/middleware.go:167]: This could be moved into S3 as a custom middleware generated there, but while its usage is specific to S3, its behavior could be used in other services that do support unsigned payload. Not sure S3.

Comment (skotambkar) [reply aws/signer/v4/middleware.go:167]: We can keep this for S3 and generate for services that use Unsignedpayload trait?

File: codegen/smithy-aws-go-codegen/src/main/java/software/amazon/smithy/aws/go/codegen/customization/S3AddPutObjectUnseekableBodyDoc.java (hunk 3 above)
Comment (skotambkar) [reply codegen/smithy-aws-go-codegen/src/main/java/software/amazon/smithy/aws/go/codegen/customization/S3AddPutObjectUnseekableBodyDoc.java:33]: >  Other customizations where order model is pre-processed is where this will be relevant.

Can't think of a use case where order of pre-processing documentation would matter.

File: aws/signer/v4/middleware.go (hunk 2 above)
Comment (jasdel) [reply aws/signer/v4/middleware.go:167]: Moving this to be generated by the relevant services is a option. 

This utility needs to be present in some form because the payload hash middleware ID nor any of the concrete middleware themselves.

File: .changes/next-release/service.s3-bugfix-1615945295096775000.json (hunk 1 above)
Comment (jasdel) [reply .changes/next-release/service.s3-bugfix-1615945295096775000.json:5]: Yeah annoucement isn't a fit either since it does not have a version bump associated with it.

given this `bugfix` is still the best type unless we also add `documentation` type.

**Review by skotambkar — APPROVED** [review]

**Review by skmcgrail — APPROVED** [review]

**Comment by Thomasdezeeuw** [comment]
It seems the documentation change was undone by https://github.com/aws/aws-sdk-go-v2/pull/787, which resulted in me hitting https://github.com/aws/aws-sdk-go-v2/issues/1108 again.
